### ISSUE TITLE
JSON-e

### ISSUE THREAD

**Issue by gregsdennis** [gregsdennis/json-everything#539]
### Describe the functionality that isn't covered by this suite.

JSON-compatible templating/transforming syntax.

https://json-e.js.org/using.html
https://github.com/json-e/json-e

### Describe alternatives you've considered.

JSONata isn't JSON-compatible.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by gregsdennis** [comment]
Notes on implementation:

- `$eval` is going to be a pain, although it does seem well-specified (unlike JSON Logic)
- `$let` seems to provide a new context for `in`.  Is referencing the external context permitted from `in`?  There aren't examples of that.
- For `$map`/`each(x)`, the `x` is a user-defined variable name.  It also supports multiple variables, but each case is well-defined.
- Why are arrays concatenated for `$mergeDeep` but not for `$merge`?
- From the `$match` example, it looks like the default sorting key for `$sort` is just the root value, but that's not specified in the `$sort` section.
- To what degree can these be nested?  Is any object not found in a `$json` directive to be interpreted as a template?
- How do I specify that I just want (e.g.) `$map` as a key?  (See https://json-e.js.org/operators.html#escaping-operators, so if you want `$$$$map`, you'd need to use `$$$$$map` in the template. 🤦)
- Since directives start with `$` what happens if a directive isn't recognized, e.g. `$foo`?
- What happens if a directive doesn't get the kind of value it's expecting, e.g. `$reverse` doesn't get an array?

**Comment by djmitche** [comment]
>     * `$eval` is going to be a pain, although it does seem well-specified (unlike JSON Logic)

I'm curious why `$eval` would be a pain? It's one of two gateways to the expression language (the other being string interpolation), so it's pretty central to the whole thing!

>     * `$let` seems to provide a new context for `in`.  Is referencing the external context permitted from `in`?  There aren't examples of that.

Contexts nest. The words "amended" and "added" are intended to indicate this, but the docs could certainly be clearer.

>     * Why are arrays concatenated for `$mergeDeep` but not for `$merge`?

The "deep" merging looks "into" objects and combines them, so it should also do so for arrays. The most natural way to do so is to append. In a practical sense this usually feels natural, such as concatenating lists of ports or command-line arguments, whereas replacing one array with another does not.

>     * From the `$match` example, it looks like the default sorting key for `$sort` is just the root value, but that's not specified in the `$sort` section.

"The by(var) property defaults to the identity function."
 
>     * For expression operators, what sort of type matching is required?  Is there any type coercion, e.g. `"4.5" > 3` or `"3.0" == 3?  What happens when types don't match? (Looks like [no type coercion](https://github.com/json-e/json-e/pull/241).)

JSON treats numbers as a single type, so yes, `4.5 > 3`. I don't think it's entirely specified for corner cases, e.g., `-0.0 == -0` ([ask me how I know](https://chromium-review.googlesource.com/c/chromium/src/+/4873600)). Among the JSON types, we want to avoid implicit conversion, but the `number` function does allow explicit conversion.

I think operators over dissimilar type are forbidden, and IIRC we have tests for those in specification.yml, but if not, plz add!

>     * For the `in` operator, does it check object _values_ as well, or only keys, e.g. `"foo" in {a: 'foo', b: 'bar'}`?

Nope.

>     * To what degree can these be nested?  Is any object not found in a `$json` directive to be interpreted as a template?

To any degree. We've tried to ensure that the complexity of JSON-e rendering is roughly linear with the size of the input, but it isn't bounded.

>     * How do I specify that I just want (e.g.) `$map` as a key?  (See https://json-e.js.org/operators.html#escaping-operators, so if you want `$$$$map`, you'd need to use `$$$$$map` in the template. 🤦)

Yep. Everything is expressible, just some things less conveniently..
 
>     * Since directives start with `$` what happens if a directive isn't recognized, e.g. `$foo`?

(Note that these are called "operators") IIRC unrecognized operators are forbidden, so that we can compatibly add new operators.
 
>     * What happens if a directive doesn't get the kind of value it's expecting, e.g. `$reverse` doesn't get an array?

That's an error.

>     * If functions provided in the context are [evaluated](https://json-e.js.org/expressions.html#function-invocation) (!), does that mean that templates provided in the context are also evaluated?  Does that open up potential for injection attacks?

No, templates in the context are not evaluated.

Functions are written in the native language, and implicitly should be written to handle untrusted input. Maybe we should be explicit about that! For example, I believe Taskcluster's GitHub integration supplies a function to generate a UUID (in the form of a slugid) for a given textual input, as a way of generating unique taskIds.

>     * Several places mention raising errors, but how is that done?  Does it gracefully return with an error condition, does it throw an exception, or is does the implementation get to decide?

It's language-specific, so e.g., exceptions in Python and JS, Result::Err in Rust, and (_, err) in Go. Note that the languages are pretty inconsistent about errors! Most importantly, there are likely entire classes of errors detected in some languages and allowed to pass unnoticed in others. But also, the error messages produced are almost never the same across languages. There was some work begun to clean this up, and `specification.yml` does include some error messages, but for the most part it just asserts that there was an error for any disallowed input.

> ### Found in issues
> 
>     * Looks like https://github.com/json-e/json-e/blob/main/specification.yml is the "test suite."

Yep.

>     * [Data can be interpreted as functions!](https://github.com/json-e/json-e/issues/158#issuecomment-334585277).

More precisely, native-language functions are valid values during JSON-e rendering, but it is an error to include such a value in the result of the rendering, so as @jonasfj said `{"$eval":"min"}` is invalid. And it's an error to do much of anything other than assign and apply those values. So `min + max` is an error, `number > 2` is an error, etc.

>     * `now` is just a context value that can be [overridden](https://github.com/json-e/json-e/issues/210).

This is probably worth documenting. In the uses to which we've seen JSON-e put, `$fromNow` is very useful, so it makes sense to have a special case for `now`.

>     * The context can [provide _custom_ functions](https://github.com/json-e/json-e/issues/234#issuecomment-376216280)!

Yep! If necessary, a language implementation could omit this capability, but in practice it's just a way for a particular use of JSON-e to include some additional, purpose-specific functionality. Within JSON-e, the use of those functions is pretty constrained as described above. Critically, you can't actually _define_ a function in JSON-e.

**Comment by gregsdennis** [comment]
Thanks for the responses!

> I'm curious why `$eval` would be a pain?

I've done expression support for JSON Path and JSON Logic, and doing both in .Net is just... difficult... at times.  It's parsing, plus building up a type-safe expression model, then figuring out how to evaluate it... It's a problem I have to deal with, though, not anything about JSON-e.

> Contexts nest.

I'd be grateful for more information about how this works.  Is it just an internal stack of contexts, or is one context actually nested under some well-known key of another context?

> The "deep" merging looks "into" objects and combines them, so it should also do so for arrays. The most natural way to do so is to append. In a practical sense this usually feels natural, such as concatenating lists of ports or command-line arguments, whereas replacing one array with another does not.

Okay, so that leads to the question: why does `$merge` _not_ concat arrays?

**Comment by gregsdennis** [comment]
Thanks for the responses, @djmitche!

> I'm curious why `$eval` would be a pain?

I've done expression support for JSON Path, and doing it in .Net was just... difficult... at times.  It's parsing, plus building up a type-safe expression model, then figuring out how to evaluate it... It's a problem I have to deal with, though, not anything about JSON-e.

> Contexts nest.

I'd be grateful for more information about how this works.  Is it just an internal stack of contexts, or is one context actually nested under some well-known key of another context?

> The "deep" merging looks "into" objects and combines them, so it should also do so for arrays. The most natural way to do so is to append. In a practical sense this usually feels natural, such as concatenating lists of ports or command-line arguments, whereas replacing one array with another does not.

Okay, so that leads to the question: why does `$merge` _not_ concat arrays?

> JSON treats numbers as a single type, so yes, `4.5 > 3`

This isn't what I'm asking.  I asked about `"4.5" > 3`, which is a string containing a number and a number.

> I think operators over dissimilar type are forbidden...

👍

> Functions are written in the native language.

Is interoperability not a goal?  What if I set up an API that allows clients to provide contexts for a template that I've defined?

> Most importantly, there are likely entire classes of errors detected in some languages and allowed to pass unnoticed in others.

Again, this seems very un-interoperable.  I would expect (especially with a test suite in place) that you'd get the same results regardless of the runtime.

> so as @jonasfj said `{"$eval":"min"}` is invalid.

But in the linked comment, it was used in a `$let` statement where it was ultimately used _as a function_ meaning that in that context it was valid.  Because of this, I would need to consider that a value may be a function that will be used later.  If it's not used later, _then_ I need to raise an error.

**Comment by djmitche** [comment]
> > Contexts nest.
> 
> I'd be grateful for more information about how this works. Is it just an internal stack of contexts, or is one context actually nested under some well-known key of another context?

Yep, it's a stack of {string: value} maps, with variable lookup searching through them from the innermost to outermost.

> > The "deep" merging looks "into" objects and combines them, so it should also do so for arrays. The most natural way to do so is to append. In a practical sense this usually feels natural, such as concatenating lists of ports or command-line arguments, whereas replacing one array with another does not.
> 
> Okay, so that leads to the question: why does `$merge` _not_ concat arrays?

Because it only operates at the top level, and on objects, so `{$merge: [[1, 2], [3, 4]]}` is forbidden.

Probably the design error was to call both of them "merge" - they're doing different things!

> > Functions are written in the native language.
> 
> Is interoperability not a goal? What if I set up an API that allows clients to provide contexts for a template that I've defined?

Interop is a goal, although more on that below. But I don't think this precludes interoperability, as JSON-e is typically used in a specific situation and that situation will define where the template and context come from, and what values are in the context -- possibly including functions specific to the situation.

As an example, [here](https://docs.taskcluster.net/docs/reference/core/hooks/firing-hooks#json-e-rendering) is one use of JSON-e in Taskcluster. It defines the context (which, in some cases, contains un-trusted data from an API call, as you suggest) and how the hook's template is rendered. There are no custom functions in this particular case.

Another example is [taskcluster-github](https://docs.taskcluster.net/docs/reference/integrations/github/taskcluster-yml-v1#json-e-rendering) which provides a similar definition, but also includes a useful `as_slugid` function. In this case, there's only one implementation of the rendering, so there's no interop concerns, but if there were (for example, if some of taskcluster-github was written in JS and some was in Python) then we would need to implement `as_slugid` in both of those languages and use testing to ensure the implementations match.

> > Most importantly, there are likely entire classes of errors detected in some languages and allowed to pass unnoticed in others.
> 
> Again, this seems very un-interoperable. I would expect (especially with a test suite in place) that you'd get the same results regardless of the runtime.

Yeah, this is sort of an artifact of history. JSON-e was originally developed as a mini-language within Taskcluster, and only implemented in JS. It soon became useful to use JSON-e in other languages (I think Python was second?), but still within Taskcluster -- so a precise "spec" with precisely matching errors were not something we thought very hard about. Once we saw that JSON-e was broadly useful, this omission became more obvious, but it's quite challenging to fix retroactively and nobody has had the time to do so.

An input that are valid in some implementations and cause an error in another implementation is definitely a bug and we would fix it. We have tried to include lots of edge cases in `specification.yml` to detect any such inconsistencies. But because the language spec is not especially precise and the implementations are all quite different, I think it's quite likely that such bugs remain hidden in some cases we have failed to imagine.

For error messages (where two implementations both consider the same input invalid, but produce different error messages), the spec doesn't say anything about them. That these match seems less critical than that non-error outputs match. And, making messages match can be quite difficult where different languages approach implementation in different ways. So, this has stalled due to lack of engineering bandwidth.

> > so as @jonasfj said `{"$eval":"min"}` is invalid.
> 
> But in the linked comment, it was used in a `$let` statement where it was ultimately used _as a function_ meaning that in that context it was valid. Because of this, I would need to consider that a value may be a function that will be used later. If it's not used later, _then_ I need to raise an error.

That's correct: a value may be a function. In fact, in the Rust implementation (the most strongly-typed implementation) this meant we were unable to use the JSON library's `Value` type (`serde_json` is the JSON library), because it did not include functions. In fact, we also needed a deletion marker, so defining an internal representation made sense anyway.

https://github.com/json-e/json-e/blob/ad09d39f3ed8af54e4586eb56e444e4351744bbd/rs/src/value.rs#L47-L62

One tiny adjustment: it's not an error to not use the function. It's just an error for the return value of `render()` to contain a function (since that return value must be valid JSON). So `{$let: {f: min}, in: {$eval: 'f(4,1)'}}` is legal (using the function), `{$let: {f: min}, in: {$eval: '4 + 1'}}` is legal (not using the function), but `{$let: {f: min}, in: {$eval: 'f'}}`` is not legal (returning the function from the render),  

In the Rust implementation, all of the rendering is handled with the internal JSON-e `Value` type, and this is converted back to `serde_json::Value` once the rendering is complete. This is a recursive process, and if it hits a `Value::Function`, it errors out:

https://github.com/json-e/json-e/blob/ad09d39f3ed8af54e4586eb56e444e4351744bbd/rs/src/value.rs#L210-L235

**Comment by gregsdennis** [comment]
Thanks for all that!  It really helps.

**Comment by gregsdennis** [comment]
@djmitche more of a casual question:

Looking at [escaping operators](https://json-e.js.org/operators.html#escaping-operators), have you seen anyone cascade evaluations so that the first eval removes an escape, and then the result is subsequently evaluated to get another result?

**Comment by djmitche** [comment]
Yes, and generally it's led to massive confusion and I don't recommend it.

**Comment by gregsdennis** [comment]
Interesting... _[a mental door opens...]_ 😈

**Comment by gregsdennis** [comment]
Of just over 1000 tests, I'm down to 237 that are failing.  I still need to implement:

- string interpolation
- functions as values
- operator escaping
- `$map`

After that, it's just basic debugging.

**Comment by djmitche** [comment]
That's fantastic!

**Comment by gregsdennis** [comment]
Found an issue with the YAML parser I'm using.  [Reported](https://github.com/aaubry/YamlDotNet/issues/867).  Hopefully they fix it.

**Comment by gregsdennis** [comment]
I've updated my tests to just ensure that an error is thrown when expected.  If the message is different, it results as inconclusive, which isn't a failure.

As of now, I only have 31 failures, and they're all pretty much literal parsing issues, which I've been avoiding.

**Comment by gregsdennis** [comment]
Down to 10!  I've had to considerably overhaul my expression parsing to get this.

![image](https://github.com/gregsdennis/json-everything/assets/2676804/e0c3ab60-af49-4dc6-9398-9224f81f4298)
